### PR TITLE
Implement socket_vmnet network (QEMU)

### PR DESCRIPTION
**Pull request by spowelljr**
Closes https://github.com/kubernetes/minikube/issues/15038
Closes https://github.com/kubernetes/minikube/issues/14692

Implemented socket_vmnet network for the QEMU driver

**minikube start:**
```
$ minikube start --driver qemu --network socket_vmnet
😄  minikube v1.27.0 on Darwin 12.6 (arm64)
✨  Using the qemu2 (experimental) driver based on user configuration
❗  Using qemu with 'socket_vmnet' network is experimental
👍  Starting control plane node minikube in cluster minikube
🔥  Creating qemu2 VM (CPUs=2, Memory=4000MB, Disk=20000MB) ...
🐳  Preparing Kubernetes v1.25.1 on Docker 20.10.18 ...
    ▪ Generating certificates and keys ...
    ▪ Booting up control plane ...
    ▪ Configuring RBAC rules ...
🔎  Verifying Kubernetes components...
    ▪ Using image gcr.io/k8s-minikube/storage-provisioner:v5
🌟  Enabled addons: default-storageclass, storage-provisioner
🏄  Done! kubectl is now configured to use "minikube" cluster and "default" namespace by default
```

**minikube service:**
```
$ kubectl create deployment hello-minikube --image=docker.io/nginx:1.23
$ kubectl expose deployment hello-minikube --type=NodePort --port=80
$ minikube service hello-minikube
|-----------|----------------|-------------|----------------------------|
| NAMESPACE |      NAME      | TARGET PORT |            URL             |
|-----------|----------------|-------------|----------------------------|
| default   | hello-minikube |          80 | http://192.168.105.2:32037 |
|-----------|----------------|-------------|----------------------------|
🎉  Opening service default/hello-minikube in default browser...
$ curl 192.168.105.2:32037
<!DOCTYPE html>
<html>
<head>
<title>Welcome to nginx!</title>
...
```

**minikube tunnel:**
```
$ kubectl create deployment balanced --image=docker.io/nginx:1.23
$ kubectl expose deployment balanced --type=LoadBalancer --port=80
$ minikube tunnel

# in new terminal window
$ kubectl get services balanced
NAME       TYPE           CLUSTER-IP     EXTERNAL-IP    PORT(S)        AGE
balanced   LoadBalancer   10.101.76.73   10.101.76.73   80:30925/TCP   22s
$ curl 10.101.76.73:80
<!DOCTYPE html>
<html>
<head>
<title>Welcome to nginx!</title>
...
```

**starting with user network (backwards-compatibility)**
```
$ minikube start --driver qemu
😄  minikube v1.27.0 on Darwin 12.6 (arm64)
✨  Using the qemu2 (experimental) driver based on user configuration
❗  The default network for QEMU will change from 'user' to 'socket_vmnet' in a future release
❗  You are using the QEMU driver without a dedicated network, which doesn't support `minikube service` & `minikube tunnel` commands.
To try the experimental dedicated network see: https://minikube.sigs.k8s.io/docs/drivers/qemu/#networking
👍  Starting control plane node minikube in cluster minikube
🔥  Creating qemu2 VM (CPUs=2, Memory=4000MB, Disk=20000MB) ...
🐳  Preparing Kubernetes v1.25.1 on Docker 20.10.18 ...
    ▪ Generating certificates and keys ...
    ▪ Booting up control plane ...
    ▪ Configuring RBAC rules ...
🔎  Verifying Kubernetes components...
    ▪ Using image gcr.io/k8s-minikube/storage-provisioner:v5
🌟  Enabled addons: storage-provisioner, default-storageclass
🏄  Done! kubectl is now configured to use "minikube" cluster and "default" namespace by default
```

### CODE REVIEW

File: cmd/minikube/cmd/start_flags.go [hunk 1]
```diff
@@ -457,6 +457,18 @@ func getCNIConfig(cmd *cobra.Command) string {
 	return chosenCNI
 }
 
+func getNetwork(driverName string) string {
+	n := viper.GetString(network)
+	if n != "" {
+		return n
+	}
+	if driver.IsQEMU(driverName) {
+		out.WarningT("The default network for QEMU will change from 'user' to 'socket' in a future release")
```
Comment (medyagh) [inline cmd/minikube/cmd/start_flags.go:466]: lets add a website entry for scoket_vmnet
and talk about how it is different and how to install it

also lets add this to the message

"You are using qemu driver without a dedicated network, which does not have support for some networking commands such as 'minikbue service
 or 'minikube tunnel' if you like to try the experimental dedicated driver checkout the instructions on ...URL..."

**Review by medyagh — CHANGES_REQUESTED** [review]
the PR looks good to me except we need a documentation page, and also I think we might wanna change the name "socket" to be something more meaningful

I hate to use overloaded words, socket means a lot of things in many worlds.

how about we use a name like "socket_vmnet" or "lima_socket" ?

I also like to know what @afbjorklund thinks

File: cmd/minikube/cmd/start_flags.go (hunk 1 above)
Comment (spowelljr) [reply cmd/minikube/cmd/start_flags.go]: Added a website entry describing the two different network options and how to install socket_vmnet.

Also added the CLI message to warn users when they're starting QEMU without a dedicated network.

File: site/content/en/docs/drivers/qemu.md [hunk 2]
```diff
@@ -29,6 +29,33 @@ minikube start supports some qemu specific flags:
   * Note: while the flag should override the config, if the flag does not take effect try running `minikube delete`.
   * MacPorts: if you are installing [minikube](https://ports.macports.org/port/minikube/) and [qemu](https://ports.macports.org/port/qemu/) via MacPorts on a Mac with M1, use the following flag: `--qemu-firmware-path=/opt/local/share/qemu/edk2-aarch64-code.fd`
 
+## Networking
+
+The QEMU driver has two networking options, `user` & `socket_vmnet`.
+
+{{% tabs %}}
+{{% tab "user - limited functionality" %}}
+The `user` network is not a dedicated network, it doesn't support some networking commands such as `minikube service` and `minikube tunnel`.
```
Comment (AkihiroSuda) [inline site/content/en/docs/drivers/qemu.md:38]:
```suggestion
The `user` network is not a dedicated network, it doesn't support some networking commands such as `minikube service` and `minikube tunnel`, and its IP address is not reachable from the host.
```

Comment (spowelljr) [reply site/content/en/docs/drivers/qemu.md:38]: Thanks for the added information @AkihiroSuda

File: cmd/minikube/cmd/start_flags.go [hunk 3]
```diff
@@ -457,6 +457,18 @@ func getCNIConfig(cmd *cobra.Command) string {
...
+		return "user"
+	}
+	return n
```
[the remark this answers is not in the document]
Comment (spowelljr) [reply cmd/minikube/cmd/start_flags.go:472]: I believe this comment is on an old commit, could you recheck this and verify that it's fixed now.

File: pkg/drivers/qemu/qemu.go [hunk 4]
```diff
@@ -465,9 +458,40 @@ func (d *Driver) Start() error {
...
+			return nil
+		}
+		// Implement a retry loop because IP address isn't added to dhcp leases file immediately
+		for i := 0; i < 30; i++ {
+			log.Debugf("Attempt %d", i)
+			err = getIP()
+			if err == nil {
+				break
+			}
+			time.Sleep(2 * time.Second)
```
[the remark this answers is not in the document]
Comment (spowelljr) [reply pkg/drivers/qemu/qemu.go:483]: My concern with exponential backoff is that is if it takes 15 seconds for the entry to populate we'll be waiting until the 30 second mark for the next check, making the user wait 15 seconds unnecessarily. Where in the current code the user never has to wait more than 2 seconds after the entry populates.

The check is quite a quick operation as well as it's simply reading the `/var/db/dhcpd_leases` so I don't see a problem doing the read every 2 seconds. If this was a network/API call I could definitely see the advantage to exponential backoff.

That's my reasoning for leaving it as is, however I'm happy to hear reasonings to move to exponential backoff so please let me know if I'm overlooking anything.

**Review by medyagh — APPROVED** [review]

File: pkg/drivers/common.go [hunk 5]
```diff
@@ -33,6 +37,15 @@ import (
 	"k8s.io/minikube/pkg/util"
 )
 
+const (
```
Comment (klaases) [inline pkg/drivers/common.go:40]: Inline single entries.

File: pkg/drivers/common.go [hunk 6]
```diff
@@ -33,6 +37,15 @@ import (
...
+	// LeasesPath is the path to dhcpd leases
+	LeasesPath = "/var/db/dhcpd_leases"
+)
+
+var (
```
Comment (klaases) [inline pkg/drivers/common.go:45]: Inline single entries.

File: pkg/drivers/common.go (hunk 5 above)
Comment (spowelljr) [reply pkg/drivers/common.go]: Done

File: pkg/drivers/common.go (hunk 6 above)
Comment (spowelljr) [reply pkg/drivers/common.go]: Done

**Review by klaases — APPROVED** [review]
Nice work, well done!

**Comment by Hamidasl** [comment]
https://github.com/users/SashaITx/achievements/pull-sharksources.zip

File: cmd/minikube/cmd/docker-env.go [hunk 7]
```diff
@@ -50,6 +50,7 @@ import (
 	"k8s.io/minikube/pkg/minikube/reason"
```
Comment (Hamidasl) [inline cmd/minikube/cmd/docker-env.go:50]: 8c12c40853390f6efa8da0aef40dc1f6b64e32dd

**Pull request merged**